File: Linear Data Structures/solutions/Queue.py
class Node:
    def __init__(self, val=None, next=None):
        self.val = val
        self.next = next

class Queue:
    # Note: Utilize the tail pointer to achieve O(1) enqueue
    # time complexity
    def __init__(self, head=None, tail=None):
        self.head = head
        self.tail = tail

    # Insert a value into a queue
    def enqueue(self, n):
        if self.head is None:
            self.head = Node(n)
            self.tail = self.head
            return
        
        newNode = Node(n)
        self.tail.next = newNode
        self.tail = self.tail.next

    # Remove a value from a queue
    def dequeue(self):
        if self.head is None:
            return "Can't dequeue. Queue is empty."

        temp = self.head.next
        del self.head
        
        self.head = temp

    # Check if queue is empty
    def isEmpty(self):
        if self.head is not None:
            return False

        return True

    # Output all of the value(s) in queue
    def __str__(self):
        curr = self.head
        s = ""

        while curr:
            s += str(curr.val) + " "
            curr = curr.next

        return s

File: Linear Data Structures/solutions/Java/Queue.java
/*
 * This is a Queue Data Structure Implementation Using LinkedList
 */
public class Queue {

    static class Node {
        Object value;
        Node next;

        Node (Object value) {
            this.value = value;
        }
    }

    private Node head;
    private Node tail;

    // Constructor
    /*
     * Utilize the tail pointer to achieve O(1) time complexity
     */
    public Queue () {
        this.head = null;
        this.tail = null;
    }

    /*
     * Inserting a value into the queue
     */
    public void enqueue (Object n) {
        // Code here
        Node newNode = new Node(n);

        if (this.head == null) {
            this.head = newNode;
            this.tail = this.head;
        } else {
            this.tail.next = newNode;
            this.tail = this.tail.next;
        }
    }

    /* 
     * Removing a value from the queue using FIFO
     */
    public void dequeue () {
        // Code here
        if (this.head == null) {
            System.out.println("Nothing to dequeue. Queue is empty.");
        } else {
            this.head = this.head.next;
        }
    }

    /*
     * Check if the Queue is empty
     * @return True if the queue is empty, else false
     */
    public boolean isEmpty() {
        if (this.head == null) {
            return true;
        }
        
        return false; // Just a temporary placeholder
    }

    @Override
    /*
     * Printing out all the value in the queue
     */
    public String toString() {
        Node curr = this.head;
        String s = "";

        while (curr != null) {
            s += curr.value + " ";
            curr = curr.next;
        }

        return s;
    }
}
